[DataPipe] Adding serialization test at different stages of reading for IterDataPipes

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73119
* #73118
* #72896

Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

Differential Revision: [D34354496](https://our.internmc.facebook.com/intern/diff/D34354496)

NivekT committed:
Adding serialization test at different stages of reading for IterDataPipes

[ghstack-poisoned]

NivekT committed:
Update on "Adding serialization test at different stages of reading for IterDataPipes"


Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

[ghstack-poisoned]

ejguan commented on test/test_datapipe.py:
```diff
@@ -562,6 +562,15 @@ def _worker_init_fn(worker_id):
 class TestFunctionalIterDataPipe(TestCase):
 
+    def _serialization_test_helper(self, datapipe, has_two_children=False):
+        serialized_dp = pickle.dumps(datapipe)
+        deserialized_dp = pickle.loads(serialized_dp)
```
nit: `deserialized_dp = pickle.loads(pickle.dumps(datapipe))`

NivekT replied:
Would it be good to keep them separate so that it will be more obvious whether `dumps` or `loads` throws an error when one of them fails?

ejguan replied:
But, you have already tested it before invoking them. See comment below.

ejguan commented on test/test_datapipe.py:
```diff
@@ -589,19 +598,40 @@ def test_serializable(self):
 
     def test_serializable_with_dill(self):
-        """Only for DataPipes that take in a function or buffer as argument"""
+        """Only for DataPipes that take in a function as argument"""
```
Can we do the same thing for `dill`?

NivekT replied:
As in testing serialization at different stages of reading?

NivekT replied:
This is done

NivekT committed:
Update on "Adding serialization test at different stages of reading for IterDataPipes"


Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354496).

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -589,19 +598,40 @@ def test_serializable(self):
         dp_compare_children = {dp.iter.Demultiplexer, dp.iter.Forker}
+
         for dpipe, dp_args, dp_kwargs in picklable_datapipes:
             datapipe = dpipe(input_dp, *dp_args, **dp_kwargs)  # type: ignore[call-arg]
```
Better move in into the if branch and initialize as dp1, dp2 = dpipe(...) in case of dp_skip_comparison, it is going to be less confusing this way.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -589,19 +598,40 @@ def test_serializable(self):
+                else:  # DataPipes that need to compare children
+                    it1, it2 = iter(datapipe[0]), iter(datapipe[1])
+                    _, _ = next(it1), next(it2)
+                    self._serialization_test_helper(datapipe, has_two_children=True)
```
You can drop has_two_children and just call _serialization_test_helper twice for dp1 and dp2

ejguan commented on test/test_datapipe.py:
```diff
@@ -589,19 +598,40 @@ def test_serializable(self):
             datapipe = dpipe(input_dp, *dp_args, **dp_kwargs)  # type: ignore[call-arg]
-            serialized_dp = pickle.dumps(datapipe)
-            deserialized_dp = pickle.loads(serialized_dp)
+            serialized_dp = pickle.dumps(datapipe)  # Making sure dpipe is picklable
+            _ = pickle.loads(serialized_dp)  # Making sure dpipe is loadable
```
You can remove these two lines.

NivekT replied:
I have refactored the code to make it clear what I'm doing

NivekT committed:
Update on "[DataPipe] Adding serialization test at different stages of reading for IterDataPipes"


Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

Differential Revision: [D34354496](https://our.internmc.facebook.com/intern/diff/D34354496)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354496).

NivekT committed:
Update on "[DataPipe] Adding serialization test at different stages of reading for IterDataPipes"


Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

Differential Revision: [D34354496](https://our.internmc.facebook.com/intern/diff/D34354496)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding serialization test at different stages of reading for IterDataPipes"


Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

Differential Revision: [D34354496](https://our.internmc.facebook.com/intern/diff/D34354496)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding serialization test at different stages of reading for IterDataPipes"


Test if a DataPipe is serializable after its contents are partially read and completely read. This is especially important for DataPipes with buffers.

Differential Revision: [D34354496](https://our.internmc.facebook.com/intern/diff/D34354496)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354496).
